### DISCUSSION TITLE
R - Find/replace in dataframe using vectors of strings

### DISCUSSION THREAD

**Discussion by kubu4** [RobertsLab/resources#1242] [Q&A] [answered]
I have a dataframe which looks like this:

```
     GOslim                                   Term
 GO:0000003                           reproduction
 GO:0000278                     mitotic cell cycle
 GO:0000902                     cell morphogenesis
 GO:0002376                  immune system process
 GO:0003013             circulatory system process
 GO:0005975         carbohydrate metabolic process
 GO:0006091 generation of precursor metabolites...
 GO:0006259                  DNA metabolic process
 GO:0006397                        mRNA processing
 GO:0006399                 tRNA metabolic process
 GO:0006412                            translation
 GO:0006457                        protein folding
 GO:0006464 cellular protein modification proce...
 GO:0006520 cellular amino acid metabolic proce...
 GO:0006605                      protein targeting
 GO:0006629                lipid metabolic process
 GO:0006790      sulfur compound metabolic process
 GO:0006810                              transport
 GO:0006913            nucleocytoplasmic transport
 GO:0006914                              autophagy
 GO:0006950                     response to stress
 GO:0007005             mitochondrion organization
 GO:0007009           plasma membrane organization
 GO:0007010              cytoskeleton organization
 GO:0007034                     vacuolar transport
 GO:0007049                             cell cycle
 GO:0007059                 chromosome segregation
 GO:0007155                          cell adhesion
 GO:0007165                    signal transduction
 GO:0007267                    cell-cell signaling
 GO:0007568                                  aging
 GO:0008150                     biological_process
 GO:0008219                             cell death
 GO:0008283          cell population proliferation
 GO:0009056                      catabolic process
 GO:0009058                   biosynthetic process
 GO:0009790                     embryo development
 GO:0015031                      protein transport
 GO:0015979                         photosynthesis
 GO:0016192             vesicle-mediated transport
 GO:0019748            secondary metabolic process
 GO:0021700               developmental maturation
 GO:0022607            cellular component assembly
 GO:0022618     ribonucleoprotein complex assembly
 GO:0030154                   cell differentiation
 GO:0030198      extracellular matrix organization
 GO:0030705 cytoskeleton-dependent intracellula...
 GO:0032196                          transposition
 GO:0034330             cell junction organization
 GO:0034641 cellular nitrogen compound metaboli...
 GO:0034655 nucleobase-containing compound cata...
 GO:0040007                                 growth
 GO:0040011                             locomotion
 GO:0042254                    ribosome biogenesis
 GO:0042592                    homeostatic process
 GO:0043473                           pigmentation
 GO:0044281       small molecule metabolic process
 GO:0044403                      symbiotic process
 GO:0048646 anatomical structure formation invo...
 GO:0048856       anatomical structure development
 GO:0048870                          cell motility
 GO:0050877                 nervous system process
 GO:0051276                chromosome organization
 GO:0051301                          cell division
 GO:0051604                     protein maturation
 GO:0055085                transmembrane transport
 GO:0061024                  membrane organization
 GO:0065003    protein-containing complex assembly
 GO:0071554 cell wall organization or biogenesi...
 GO:0071941       nitrogen cycle metabolic process
 GO:0140014               mitotic nuclear division
```

I want to replace all truncated terms (e.g. `generation of precursor metabolites...`) with the full description (e.g. `generation of precursor metabolites and energy`).

Since there are multiple truncated terms, I'd like to replace them all with the appropriate full description. The catch is, I'd like to do it elegantly. I can do each individually like so:

```R
df %>% mutate_all(funs(str_replace(., "generation of precursor metabolites...",
                                           "generation of precursor metabolites and energy")))
```

But, I don't want to have to repeat that over in over, where I swap out the find/replace terms. 

I'd prefer an option where I have two vectors: 
```R
original_term <- c()

full_term <- c()
```

Then, I somehow magically have R use those to do the find/replace.

**Comment by sr320** [comment]
It is not clear if you already have the full terms in a dataframe. If you already do why not just join?

**Reply by kubu4** [reply]
The terms appear as shown in the dataframe with the trailing `...`. This is a product of the `GSEAbase` package.

I suppose I could modify the source code to _not_ spit out truncated terms. I think `GSEAbase` has something in the code that sets a character limit for long descriptions or something like that.

**Reply by laurahspencer** [reply]
so, the full description is contained in the R object, you just aren't happy with how it reads out to the console / RStudio?  How will you use the full description?  For instance, if you're printing to an RMarkdown file you might be able to modify the number of characters that get truncated ... https://stackoverflow.com/questions/36800475/avoid-string-printed-to-console-getting-truncated-in-rstudio

**Reply by kubu4** [reply]
> so, the full description is contained in the R object

I'm not certain. I looked into this annoyance awhile ago and vaguely remember seeing code that specifically replaces the end of a string with `...` if the number of characters in the string exceeds a set number. So, the full description is likely _not_ in the object, as it gets processed when the object is being created.

At the time, and now, I just decided it would be easier to perform a find/replace after I had my dataframe.

**Reply by kubu4** [reply]
Found where it is in the source code. Previously was addressing some of this GSEAbase stuff in [this GitHub issue](https://github.com/RobertsLab/resources/issues/842), in which I linked to the source code.

[Here's a link to the source code](https://github.com/Bioconductor/GSEABase/blob/master/R/methods-OBOCollection.R).

And, here's the relevant section (lines 266 - 274) where the code truncates the GOslim descriptions:

```R
    ## Prepare a data frame for results
    term <- vapply(terms, Term, character(1), USE.NAMES=FALSE)
    trunc <- ifelse(nchar(term) > 35, "...", "")
    df <- data.frame(
        Slim=names(terms),
        Count=0L, Percent=0,
        Term=sprintf("%.35s%s", term, trunc),
        row.names=1
    )
```

Other than altering the source code to allow for more characters (or, just eliminating the truncation), anyone know how to tackle this in a fancy R way, without the need for find/replace?

**Comment by kubu4** [comment]
Also, on a side note, I came across some info that [indicates TopGO does the same truncation, but can be remedied by adding the `numChars=1000` argument to the `GenTable()` function](https://support.bioconductor.org/p/131412/). I know some people have talked about TopGO recently, so figured I'd just drop this here, too, since it's related.

**Comment by kubu4** [comment, marked as answer]
I'm still curious about the find/replace solution, however, the [the developer of GSEAbase has fixed the problem](https://github.com/Bioconductor/GSEABase/issues/5#issuecomment-866401489)! That was really fast and really nice! If anyone's interested in getting the fix, you'll need to install the developer version of GSEAbase (follow the link for instructions on how to do this).
